posydon/utils/posydonwarning.py

```diff
@@ -77,33 +105,336 @@ def __init__(self, message=''):
+                    else:
+                        raise TypeError(f"{attr} has to be a "
+                                        f"{type(getattr(self, attr))}.")
+                else:
+                    raise AttributeError(f"{attr} unknown to "
+                                         "_Catched_POSYDON_Warnings.")
+            if ((self.catch_warnings==False) and
+                (len(self.catched_warnings)>0)):
+                # If there are recorded warnings issue them and empty the list
+                for w in self.catched_warnings:
+                    w["stacklevel"] += 2
+                    if self.filter_first:
+                        warnings.warn(**w)
+                    else:
+                        _issue_warn(w)
+                self.catched_warnings = []
+        if new_warning is not None:
+            # Process new warning
+            if self.catch_warnings:
+                # Catch warning
+                if self.filter_first:
+                    # Apply POSYDON filtering/stats
+                    new_warning = _apply_POSYDON_filter(new_warning)
+                if self.record and new_warning is not None:
+                    # Add warning to catched ones
+                    self.catched_warnings.append(new_warning)
+            else:
+                # Issue warning
+                new_warning["stacklevel"] += 2
+                _issue_warn(new_warning)
+    
+    def __del__(self):
+        """Destructor of the object. It will issue still recorded warnings."""
+        if len(self.catched_warnings)>0:
+            # If there are recorded warnings issue them.
+            self.catch_warnings = False
+            print("There are still recorded warnings:")
+            for w in self.catched_warnings:
+                w["stacklevel"] = 2
+                if self.filter_first:
+                    warnings.warn(**w)
+                else:
+                    _issue_warn(w)
+    
+    def got_called(self):
+        """Returns, whether the object got called."""
+        return self._got_called
+    
+    def has_records(self):
+        """Checks whether there are records of catched warnings.
+
+        Returns
+        -------
+        True if there are recorded warnings otherwise False.
+        """
+        return len(self.catched_warnings)>0
+    
+    def get_cache(self, empty_cache=False):
+        """Get catched warnings.
+
+        Parameters
+        ----------
+        empty_cache : bool (default: False)
+            If True, the catched warnings will be reset.
+
+        Returns
+        -------
+        List of recorded warnings.
+        """
+        cache = copy.copy(self.catched_warnings)
+        if empty_cache:
+            self.reset_cache()
+        return cache
+    
+    def reset_cache(self):
+        """Resets the catched warnings."""
+        self.catched_warnings = []
+
+_CATCHED_POSYDON_WARNINGS = _Catched_POSYDON_Warnings()
+
+
+class Catch_POSYDON_Warnings:
```

Comment:
I am confused, does this catch all warnings, including python warnings, or just POSYDON warnings? I think we want to keep functionality to catch both, at least for the use in binarypopulation.

Reply:
We can't catch python warnings, because they never call any of our functions. But as we said that we want to have all our Warnings to be subclasses of POSYDONwarnings, those are all caught. It's our responsibility to not call other functions with arguments causing warnings in there.

Reply:
Ok, I am thinking that in the main catch_warnings block in binarypopulation, we want to catch both, mostly because the way things are currently set-up, if _any_ warning for a binary occurs, we want to record this in the oneline dataframe. So we need to know if python warnings were caught, too. Would it make sense then to use the regular python catch_warnings there?

Reply:
In general if you have an idea of how to get the functionality in binarypopulation be consistent with what happening before using the new POSYDON warnings stuff, that would be helpful.

Reply:
Having the exact same behavior as before is not possible, without put bugs in our code.

Now, there is an option for backward compatibility to put the python catching on top, which brings all the drawbacks of the python catching.
Be aware, that future code validation, e.g. in unit testing, will require the code to work without that option, because we should handle all warnings on our own.